python/ray/tests/test_autoscaling_policy.py

```diff
@@ -0,0 +1,579 @@
+        to_allocate = []
+        if (
+            pg.strategy == PlacementStrategy.STRICT_PACK
+            or pg.strategy == PlacementStrategy.PACK
+        ):
+            combined = collections.defaultdict(float)
+            for bundle in pg.bundles:
+                for k, v in bundle.items():
+                    combined[k] += v
+            ip, node_to_run = self._get_node_to_run(combined, nodes)
+            if node_to_run is None:
+                return False
+            to_allocate.append((combined, ip))
+        elif (
+            pg.strategy == PlacementStrategy.STRICT_SPREAD
+            or pg.strategy == PlacementStrategy.SPREAD
+        ):
+            # TODO (Alex): More accurate handling of non-STRICT_PACK groups.
+            remaining_nodes = nodes.copy()
+            for bundle in pg.bundles:
+                ip, node_to_run = self._get_node_to_run(bundle, remaining_nodes)
+                if node_to_run is None:
+                    return False
+                del remaining_nodes[ip]
+                to_allocate.append((bundle, ip))
+
+        for bundle, ip in to_allocate:
+            node = self.ip_to_nodes[ip]
+            node.allocate(bundle)
+        pg.start_time = self.virtual_time
+        end_time = self.virtual_time + pg.duration
+        self.event_queue.put(
+            Event(end_time, SIMULATOR_EVENT_PG_DONE, (pg, to_allocate))
+        )
+        if pg.start_callback:
+            pg.start_callback()
+        return True
+
+    def _schedule_task(self, task, nodes):
+        ip, node = self._get_node_to_run(task.resources, nodes)
+        if node is None:
+            return False
+
+        node.allocate(task.resources)
+        task.node = node
+        task.start_time = self.virtual_time
+        end_time = self.virtual_time + task.duration
+        self.event_queue.put(Event(end_time, SIMULATOR_EVENT_TASK_DONE, task))
+        if task.start_callback:
+            task.start_callback()
+        return True
+
+    def schedule(self):
+        # TODO (Alex): Implement a more realistic scheduling algorithm.
+        new_work_queue = []
+        for work in self.work_queue:
+            if isinstance(work, Task):
+                scheduled = self._schedule_task(work, self.ip_to_nodes)
+            elif isinstance(work, PlacementGroup):
+                scheduled = self._schedule_placement_group(
+                    work, self.ip_to_nodes
+                )
+            else:
+                assert False, "Unknown work object!"
+
+            if scheduled is False:
+                new_work_queue.append(work)
+        self.work_queue = new_work_queue
+
+    def _launch_nodes(self):
+        """Launch all queued nodes. Since this will be run serially after
+        `autoscaler.update` there are no race conditions in checking if the
+        queue is empty.
+        """
+        while not self.node_launcher.queue.empty():
+            config, count, node_type = self.node_launcher.queue.get()
```

Comment:
Is it relevant at this stage to assert conditions on the number of nodes launched here?

Reply:
i don't think so? if someone really wanted to check on the exact type of node at a given point in time, they should assert it in their test case.